trying one last time

SvenDowideit opened the pull request:
Lets just say that I enjoyed codesee's PR diagram - so it worked :)

Commit:
use the Store interface IsErrObjectNotFound() to avoid revealing internal errors

Signed-off-by: Sven Dowideit <sven.dowideit@portainer.io>

Commit:
what happens when you extract the datastore interfaces into their own package

Signed-off-by: Sven Dowideit <sven.dowideit@portainer.io>

Commit:
Start renaming Storage methods

Signed-off-by: Sven Dowideit <sven.dowideit@portainer.io>

Commit:
extract the boltdb specific code from the Portainer storage code (example, the others need the same)

Signed-off-by: Sven Dowideit <sven.dowideit@portainer.io>

Commit:
more extract bolt.Tx from datastore code

Signed-off-by: Sven Dowideit <sven.dowideit@portainer.io>

Commit:
minimise imports by putting moving the struct definition into the file that needs the Service imports

Signed-off-by: Sven Dowideit <sven.dowideit@portainer.io>

Commit:
more extraction of boltdb.Tx

Signed-off-by: Sven Dowideit <sven.dowideit@portainer.io>

Commit:
extract the use of bucket.SetSequence

Signed-off-by: Sven Dowideit <sven.dowideit@portainer.io>

Commit:
almost done - just endpoint.Synchonise :/

Signed-off-by: Sven Dowideit <sven.dowideit@portainer.io>

Commit:
so, endpoint.Synchonize looks hard, but i can't find where we use it, so 'delete first refactoring'

Signed-off-by: Sven Dowideit <sven.dowideit@portainer.io>

Commit:
fix test compile errors

Signed-off-by: Sven Dowideit <sven.dowideit@portainer.io>

Commit:
test compile fixes after rebase

Signed-off-by: Sven Dowideit <sven.dowideit@portainer.io>

Commit:
fix a mis-remembering I had wrt deserialisation - last time i used AnyData - jsoniter's bindTo looks interesting for the same reason

Signed-off-by: Sven Dowideit <sven.dowideit@portainer.io>

Commit:
set us up to make the connection an interface

Signed-off-by: Sven Dowideit <sven.dowideit@portainer.io>

Commit:
make the db connection a datastore interface, and separate out our datastore services from the bolt ones

Signed-off-by: Sven Dowideit <sven.dowideit@portainer.io>

Commit:
rename methods to something less oltdb internals specific

Signed-off-by: Sven Dowideit <sven.dowideit@portainer.io>

Commit:
these errors are not boltdb secific

Signed-off-by: Sven Dowideit <sven.dowideit@portainer.io>

Commit:
start using the db-backend factory method too

Signed-off-by: Sven Dowideit <sven.dowideit@portainer.io>

Commit:
export boltdb raw in case we can't export from the service layer

Signed-off-by: Sven Dowideit <sven.dowideit@portainer.io>

Commit:
add a raw export from boltdb to yaml for broken db's, and an export services to yaml in backup

Signed-off-by: Sven Dowideit <sven.dowideit@portainer.io>

Commit:
add the version info by hand for now

Signed-off-by: Sven Dowideit <sven.dowideit@portainer.io>

Commit:
actually, the export from services can be fully typed - its the import that needs to do more work

Signed-off-by: Sven Dowideit <sven.dowideit@portainer.io>

Commit:
redo raw export, and make import capable of using it

Signed-off-by: Sven Dowideit <sven.dowideit@portainer.io>

Commit:
add DockerHub

Signed-off-by: Sven Dowideit <sven.dowideit@portainer.io>

Commit:
migration from anything older than v1.21.0 has been broken for quite a while, deleting the un-tested code

Signed-off-by: Sven Dowideit <sven.dowideit@portainer.io>

Commit:
fix go test ./... again

Signed-off-by: Sven Dowideit <sven.dowideit@portainer.io>

Commit:
my goland wasn't setup to gofmt

Signed-off-by: Sven Dowideit <sven.dowideit@portainer.io>

Commit:
move the two extremely dubious migration tests down into store, so they can use the test store code

Signed-off-by: Sven Dowideit <sven.dowideit@portainer.io>

Commit:
the migrator is now free of boltdb

Signed-off-by: Sven Dowideit <sven.dowideit@portainer.io>

Commit:
reverse goland overzealous replcement of internal with boltdb

Signed-off-by: Sven Dowideit <sven.dowideit@portainer.io>

Commit:
more undo over-zealous goland internal->boltdb

Signed-off-by: Sven Dowideit <sven.dowideit@portainer.io>

Commit:
yay, now bolt is only mentioned inside the api/database/ dir

Signed-off-by: Sven Dowideit <sven.dowideit@portainer.io>

Commit:
and this might be the last of the boltdb references?

Signed-off-by: Sven Dowideit <sven.dowideit@portainer.io>

Commit:
add todo

Signed-off-by: Sven Dowideit <sven.dowideit@portainer.io>

Commit:
extract the store code into a separate module too

Signed-off-by: Sven Dowideit <sven.dowideit@portainer.io>

Commit:
don't need the fileService in boltdb anymore

Signed-off-by: Sven Dowideit <sven.dowideit@portainer.io>

Commit:
use IsErrObjectNotFound()

Signed-off-by: Sven Dowideit <sven.dowideit@portainer.io>

Commit:
use a string to select what database backend we use

Signed-off-by: Sven Dowideit <sven.dowideit@portainer.io>

ghost commented:
### CodeSee Review Map:
[![Review these changes using an interactive CodeSee Map](https://s3.us-east-2.amazonaws.com/maps.codesee.io/images/github/SvenDowideit/portainer/2/75af25d7/7560f58ab887ca0d6775008846d3b66a64341a87.svg)](https://app.codesee.io/r/reviews?pr=2&src=https%3A%2F%2Fgithub.com%2FSvenDowideit%2Fportainer)

Review [in an interactive map](https://app.codesee.io/r/reviews?pr=2&src=https%3A%2F%2Fgithub.com%2FSvenDowideit%2Fportainer)

View [more CodeSee Maps](https://app.codesee.io/r/maps?src=https%3A%2F%2Fgithub.com%2FSvenDowideit%2Fportainer)

#### Legend
![CodeSee Map Legend](https://app.codesee.io/Legend_v2.png)

Commit:
make isNew store an ephemeral bool that doesn't stay true after we've initialised it

Signed-off-by: Sven Dowideit <sven.dowideit@portainer.io>

Commit:
move the import.json wip to a separate file so its more obvious - we'll be using it for testing, emergency fixups, and in the next part of the store work, when we improve migrations and data model lifecycles

Signed-off-by: Sven Dowideit <sven.dowideit@portainer.io>

Commit:
undo vscode formatting html

Signed-off-by: Sven Dowideit <sven.dowideit@portainer.io>

andres-portainer committed:
feat(openamt): add feature flag for OpenAMT [INT-5] (#6049)

feat(openamt): add feature flag for OpenAMT [INT-5]

ArrisLee committed:
fix(stack): git credential got reset when updating docker stack EE-1967

cmenginnz committed:
fix(edge) EE-2027 cannot connect to edge agent with high network latency (#6064)

Co-authored-by: Simon Meng <simon.meng@portainer.io>

WaysonWei committed:
fix docker pull limit not showed to non admin (#6066)

andres-portainer committed:
fix(custom-templates): avoid creation of template if the compose file does not exist EE-1470 (#6011)

fix(custom-templates): avoid creation of template if the compose file does not exist EE-1470

WaysonWei committed:
feat(ui): add option to sync portainer with system theme EE-1788 (#5812)

* add option to sync portainer with system theme

zees-dev committed:
Unit tests for `enableFeaturesFromFlags` function (#6063)

* - exporting BoolPairs CLI func
- added tests for enableFeaturesFromFlags function

* Add a test that uses a feature flag to add change the outcome of code - and test persistence, as that's the current implementation

Signed-off-by: Sven Dowideit <sven.dowideit@portainer.io>

* Minor comment updates

Co-authored-by: Sven Dowideit <sven.dowideit@portainer.io>
Co-authored-by: Stéphane Busso <stephane.busso@gmail.com>

Commit:
fix(docker-event-display): support the exec exited event type (#5990)

Signed-off-by: Sven Dowideit <sven.dowideit@portainer.io>

cheloRydel committed:
feature(kubeconfig): Do not invalidate kubeconfig upon Portainer restarting [EE-1854] (#5905)

fhanportainer committed:
fix(k8s): fixed force redeployment info text (#6042)

ArrisLee committed:
fix(k8s): k8s deployment manifest file placeholder EE-1936

cheloRydel committed:
feat(app): create react button component [EE-1948] (#6022)

Co-authored-by: Chaim Lev-Ari <chiptus@gmail.com>

chiptus committed:
feat(app): add tooltip component [EE-2047] (#6088)

chiptus committed:
refactor(app/widgets): create widgets react components [EE-1813] (#6097)

chiptus committed:
refactor(app): wrap react with StrictMode [EE-2023] (#6075)

deviantony committed:
update README and issue template

hookenz committed:
fix(dockerhub-migration): prevent duplicate migrated dockerhub entries EE-2042 (#6083)

* fix(migration) make dockerhub registry migration idempotent EE-2042

* add missing changes to make updateDockerhubToDB32 idempotent

* add tests for bad migrations

cheloRydel committed:
feat(app): added Input components [EE-2007] (#6028)

prabhat83 committed:
update help link in sidebar and readme (#6082)

prabhat83 committed:
feat(images): allow tags when importing docker image EE-1737 (#5883)

prabhat83 committed:
fix(home): poll endpoints if one is down EE-1755 (#6006)

zees-dev committed:
feat(openapi): github workflow to generate and validate openapi spec EE-2056 (#6101)

* github workflow to generate and validate openapi spec

* updated github workflow name to remove spaces and be more explicit

* added swagger-cli globally to reduce dep installation times

* removed redundant webhook payload in GET request

* fixed edgeGroupList OAS3 response model

* updated CI pipeline to convert OAS2 to OAS3 and validate OAS3 instead

* updated pipeline name to be more explicit

* removed redundant swagger-cli dependency as we are using swagger2openapi only in github CI

* fixed bug with no validation - using swagger-cli to validate

jeanfrancoisgratton committed:
style(images): remove files filter from upload image task EE-1944

Commit:
fix(logs): extra CRs in downloading container logs EE-1973 (#6041)

Commit:
fix(sidebar): clear current endpoint if deleted [EE-873] (#6052)

chiptus committed:
feat(app): introduce button selector component [EE-2004] (#6112)

cheloRydel committed:
feat(app): Slider component [EE-2004] (#6024)

Commit:
fix(logs): strip trailing comma [EE-1957] (#5975)

Commit:
fix(settings): move custom logo url [EE-1698] (#5984)

sunportainer committed:
fix(stack): support removing duplicated stacks EE-1962 (#6068)

* fix/EE-1962/cannot-same-stack-name handle multiple names duplicate case

Co-authored-by: Eric Sun <ericsun@SG1.local>

chiptus committed:
feat(app): introduce input list component [EE-2003] (#6123)

cheloRydel committed:
feature(kubeconfig): access to all kube environment contexts from within the Portainer UI [EE-1727] (#5966)

WaysonWei committed:
change the namespace selector behavior (#5768)

connorlanigan committed:
fix(webhooks): support image names containing a port number (#4526) (#5970)

This fixes a bug where image/registry names that contain a port number were inadvertently truncated (because port numbers are specified with a colon, just like the image tag).

For example, updating an image named `registry.example.com:5000/myimage:oldtag` with the new image tag `newtag` was incorrectly transformed into `registry.example.com:newtag`

chiptus committed:
feat(app): introduce input-group component [EE-2062] (#6135)

Commit:
EE-1877: Windows command line for agent isn't the same as on Linux (#5895)

Signed-off-by: Sven Dowideit <sven.dowideit@portainer.io>

Commit:
fix(helm): fix go to top issue (#6134)

Commit:
chore(build): add a PORTAINER_FLAGS env var for yarn so I can default a password, and enable feature flags (#6116)

Signed-off-by: Sven Dowideit <sven.dowideit@portainer.io>

hookenz committed:
feat(migrations): add more logging EE-2071 (#6141)

* add stacktrace when recovering a panic

* add logging to the migrations

* use string format

* add context around why we return stacktrace

cheloRydel committed:
chore(build): allow darwin binaries download [EE-2070] (#6120)

Commit:
fix(logs): copy issues caused by extra CR (#6150)

fhanportainer committed:
fix(k8s): fixed force redeployment info text (#6113)

* fix(k8s): fixed force redeployment info text

* feat(stack): added infor text when automatic update is off.

Commit:
Revert "EE-1877: Windows command line for agent isn't the same as on Linux (#5895)" (#6159)

This reverts commit 6255e8d4b514c6d7c7ac4c804ade35dffd3ff929.

fhanportainer committed:
feat(ce): fix small issues for Highlight Business Edition feature (#6043)

* feat(logs): added orange border to export button.

* fix(ldap): fixed connectivity check radio button alignment issue

* fix(be-feature): added box shadow to limited feature border

* fix(be-feature): fixed hide internal auth toggle issue.

* feat(ldap): added isLimitedFeatureSelfContained config to ldap-custom-admin-group and test-login components

* feat(auth): moved save settings button in auth page to a component.

* feat(oauth): use saveSettingsButton component in oauth

* feat(oauth): use saveSettingsButton component in internal auth

* feat(oauth): use saveSettingsButton component in MS active directory auth

* feat(oauth): use saveSettingsButton component in ldap auth

* feat(auth): added new component to index.js

* removed css inline styles.

* feat(ad): added disable check method to ad auth.

* feat(ldap): moved save settings disable method to parent component

* removed inline styles.

* fix(ldap): fixed test login misalignment issue

* fix(ldap): pass isLdapFormValid function from page component

* fix(ldap): made the toggle button in custom admin group orange when it's limited to CE.

* fix(auth): fixed save setting button in Microsoft Active Directory auth

* fix(ldap): made the assign admin toggle bright orange

agneevX committed:
fix (ui): set correct dimensions for Apple touch image asset (#5888)

cheloRydel committed:
feat(openamt): Configuration of the OpenAMT capability [INT-6] (#6071)

Co-authored-by: Sven Dowideit <sven.dowideit@portainer.io>

cheloRydel committed:
fix(openamt): fix IsFeatureFlagEnabled, rename MPS Url to MPS Server [INT-6] (#6172)

prabhat83 committed:
feat(docker): allow docker container resource settings without restart EE-1942 (#6065)

Co-authored-by: sam <sam@allofword>
Co-authored-by: sam@gemibook <huapox@126.com>
Co-authored-by: Prabhat Khera <prabhat.khera@gmail.com>

WaysonWei committed:
fix(registry): fix order of registries in drop down menu EE-1939 (#5960)


Co-authored-by: Prabhat Khera <prabhat.khera@portainer.io>

Commit:
fix(docker-event-display): EE-1968: support (event_name)[:extra info] for all event Actions, and append it to the output details (#6092)

Signed-off-by: Sven Dowideit <sven.dowideit@portainer.io>

zees-dev committed:
feat(api-key/backend): introducing support for api-key based auth EE-978 (#6079)

* feat(access-token): Multi-auth middleware support EE-1891 (#5936)

* AnyAuth middleware initial implementation with tests

* using mux.MiddlewareFunc instead of custom definition

* removed redundant comments

* - ExtractBearerToken bouncer func made private
- changed helm token handling functionality to use jwt service to convert token to jwt string
- updated tests
- fixed helm list broken test due to missing token in request context

* rename mwCheckAuthentication -> mwCheckJWTAuthentication

* - introduce initial api-key auth support using X-API-KEY header
- added tests to validate x-api-key request header presence

* updated core mwAuthenticatedUser middleware to support multiple auth paradigms

* - simplified anyAuth middleware
- enforcing authmiddleware to implement verificationFunc interface
- created tests for middleware

* simplify bouncer

Co-authored-by: Dmitry Salakhov <to@dimasalakhov.com>

* feat(api-key): user-access-token generation endpoint EE-1889 EE-1888 EE-1895 (#6012)

* user-access-token generation endpoint

* fix comment

* - introduction of apikey service
- seperation of repository from service logic - called in handler

* fixed tests

* - fixed api key prefix
- added tests

* added another test for digest matching

* updated swagger spec for access token creation

* api key response returns raw key and struct - easing testability

* test for api key prefix length

* added another TODO to middleware

* - api-key prefix rune -> string (rune does not auto-encode when response sent back to client)
- digest -> pointer as we want to allow nil values and omit digest in responses (when nil)

* - updated apikey struct
- updated apikey service to support all common operations
- updated apikey repo
- integration of apikey service into bouncer
- added test for all apikey service functions
- boilerplate code for apikey service integration

* - user access token generation tests
- apiKeyLookup updated to support query params
- added api-key tests for query params
- added api-key tests for apiKeyLookup

* get and remove access token handlers

* get and remove access token handler tests

* - delete user deletes all associated api keys
- tests for this functionality

* removed redundant []byte cast

* automatic api-key eviction set within cache for 1 hour

* fixed bug with loop var using final value

* fixed service comment

* ignore bolt error responses

* case-insensitive query param check

* simplified query var assignment

* - added GetAPIKey func to get by unique id
- updated DeleteAPIKey func to not require user ID
- updated tests

* GenerateRandomKey helper func from github.com/gorilla/securecookie moved to codebase

* json response casing for api-keys fixed

* updating api-key will update the cache

* updated golang LRU cache

* using hashicorps golang-LRU cache for api keys

* simplified jwt check in create user access token

* fixed api-key update logic on cache miss

* Prefix generated api-keys with `ptr_` (#6067)

* prefix api-keys with 'ptr_'

* updated apikey description

* refactor

Co-authored-by: Dmitry Salakhov <to@dimasalakhov.com>

* helm list test refactor

* fixed user delete test

* reduce test nil pointer errors

* using correct http 201 created status code for token creation; updated tests

* fixed swagger doc user id path param for user access token based endpoints

* added api-key security openapi spec to existing jwt secured endpoints (#6091)

* fixed flaky test

* apikey datecreated and lastused attrs converted to unix timestamp

* feat(user): added access token datatable. (#6124)

* feat(user): added access token datatable.

* feat(tokens): only display lastUsed time when it is not the default date

* Update app/portainer/views/account/accountController.js

Co-authored-by: zees-dev <63374656+zees-dev@users.noreply.github.com>

* Update app/portainer/views/account/accountController.js

Co-authored-by: zees-dev <63374656+zees-dev@users.noreply.github.com>

* Update app/portainer/views/account/accountController.js

Co-authored-by: zees-dev <63374656+zees-dev@users.noreply.github.com>

* Update app/portainer/components/datatables/access-tokens-datatable/accessTokensDatatableController.js

Co-authored-by: zees-dev <63374656+zees-dev@users.noreply.github.com>

* Update app/portainer/services/api/userService.js

Co-authored-by: zees-dev <63374656+zees-dev@users.noreply.github.com>

* feat(improvements): proposed datatable improvements to speed up dev time (#6138)

* modal code update

* updated datatable filenames, updated controller to be default class export

* fix(access-token): code improvement.

Co-authored-by: zees-dev <63374656+zees-dev@users.noreply.github.com>

* feat(apikeys): create access token view initial implementation EE-1886 (#6129)

* CopyButton implementation

* Code component implementation

* ToolTip component migration to another folder

* TextTip component implementation - continued

* form Heading component

* Button component updated to be more dynamic

* copybutton - small size

* form control pass tip error

* texttip small text

* CreateAccessToken react feature initial implementation

* create user access token angularjs view implementation

* registration of CreateAccessToken component in AngularJS

* user token generation API request moved to angular service, method passed down instead

* consistent naming of access token operations; clustered similar code together

* any user can add access token

* create access token page routing

* moved code component to the correct location

* removed isadmin check as all functionality applicable to all users

* create access token angular view moved up a level

* fixed PR issues, updated PR

* addressed PR issues/improvements

* explicit hr for horizontal line

* fixed merge conflict storybook build breaking

* - apikey test
- cache test

* addressed testing issues:
- description validations
- remove token description link on table

* fix(api-keys): user role change evicts user keys in cache EE-2113 (#6168)

* user role change evicts user api keys in cache

* EvictUserKeyCache -> InvalidateUserKeyCache

* godoc for InvalidateUserKeyCache func

* additional test line

* disable add access token button after adding token to prevent spam

Co-authored-by: Dmitry Salakhov <to@dimasalakhov.com>
Co-authored-by: fhanportainer <79428273+fhanportainer@users.noreply.github.com>

xAt0mZ committed:
fix(k8s/ingress): ensure new ports are only added to ingress only if app is published via ingress (#6153)

* fix(k8s/ingress): ensure new ports are only added to ingress only if app is published via ingress

* refactor(k8s/ingress): removed deleted ports of ingress in a single pass

huib-portainer committed:
Update endpointItem.html (#6142)

feat(home): show cpu and ram for non local endpoints EE-2077

cheloRydel committed:
fix(react): use ctrl directive in WidgetTitle component [EE-2118] (#6181)

hookenz committed:
Revert "fix(openamt): fix IsFeatureFlagEnabled, rename MPS Url to MPS Server [INT-6] (#6172)" (#6182)

This reverts commit c2673557590ed39b94668ba09a1f957ac735ec15.

hookenz committed:
fix(openamt): fix IsFeatureFlagEnabled, rename MPS Url to MPS Server (#6185)

Co-authored-by: cheloRydel <marcelorydel26@gmail.com>

cmenginnz committed:
feat(registry) EE-806 add support for AWS ECR (#6165)

* feat(ecr) EE-806 add support for aws ecr

* feat(ecr) EE-806 fix wrong doc for Ecr Region

Co-authored-by: Simon Meng <simon.meng@portainer.io>

prabhat83 committed:
verify repositry URL from template json when coping (#6036) (#6111)

chiptus committed:
fix(environments): show kubeconfig env list in dark mode (#6156)

Commit:
fix(container): prevent user from editing the portainer container it self EE-917 (#6093)

* fix(container): prevent from editing portainer container

* fix(container): prevent from editing portainer container

* Missing kill operation

* fix(container): enhance creating stack from template

* fix(docker): prevent user from editing the portainer container itself EE-917

* fix(docker): enhance code style

* fix(container): fix issues from code review

* fix(container): enhance creating stack from template

* fix(container): some code review issues

* fix(container): disable leave network when the container is portainer

* fix(container): disable leave network when the container is portainer

sunportainer committed:
Fix(stack)/update StackUpdateGit swagger info to POST EE-2019 (#6176)

* fix/EE-2019/Fix-stackgitupdate-swagger

Co-authored-by: sunportainer <ericsun@SG1.local>

prabhat83 committed:
feat(config): add base url support EE-506 (#5999)

zees-dev committed:
fallback to depracted copy text if clipboard api not available (#6200) (#6218)

zees-dev committed:
- standard user cannot delete another users api-keys (#6208) (#6217)

- added new method to get api key by ID
- added tests

WaysonWei committed:
fix app templates symbol (#6221)

cmenginnz committed:
feat(webhook) EE-2125 send registry auth haeder when update swarms service via webhook (#6220)

* feat(webhook) EE-2125 add some helpers to registry utils

* feat(webhook) EE-2125 persist registryID when creating a webhook

* feat(webhook) EE-2125 send registry auth header when executing a webhook

* feat(webhook) EE-2125 send registryID to backend when creating a service with webhook

* feat(webhook) EE-2125 use the initial registry ID to create webhook on editing service screen

* feat(webhook) EE-2125 update webhook when update registry

* feat(webhook) EE-2125 add endpoint of update webhook

* feat(webhook) EE-2125 code cleanup

* feat(webhook) EE-2125 fix a typo

* feat(webhook) EE-2125 fix circle import issue with unit test

Co-authored-by: Simon Meng <simon.meng@portainer.io>

cheloRydel committed:
fix(kubeconfig): show kubeconfig download button for non admin users [EE-2123] (#6204)

Co-authored-by: Simon Meng <simon.meng@portainer.io>

Commit:
further rebases for webhooks

Signed-off-by: Sven Dowideit <sven.dowideit@portainer.io>
